Add sensor user journey (#2986)
viamrobotics/docs: docs/use-cases/sensor-data-query.md

<!-- markdownlint-disable-file MD001 -->

{{% expand "Need to query by date? Click here." %}}

##### Query by date

When using MQL to query your data by date or time range, you can optimize query performance by avoiding the MongoDB `$toDate` expression, using the [BSON `date` type](https://www.mongodb.com/docs/manual/reference/bson-types/#date) instead.

For example, use the following query to search by a date range in the `mongosh` shell, using the JavaScript `Date()` constructor to specify an explicit start timestamp, and use the current time as the end timestamp:

```mongodb {class="command-line" data-prompt=">"}
// Switch to sensorData database:
use sensorData

// Set desired start and end times:
const startTime = new Date('2024-02-10T19:45:07.000Z')
const endTime = new Date()

// Run query using $match:
db.readings.aggregate(
    [
        { $match: {
            time_received: {
                $gte: startTime,
                $lte: endTime }
        } }
    ] )
```

{{% /expand%}}

## Next steps

> To see sensor data in action, check out these tutorials:
> 
> {{< cards >}}
> {{% card link="/use-cases/sensor-data-visualize/" %}}
> {{% card link="/tutorials/control/air-quality-fleet/" %}}
> {{< /cards >}}

Author: ```suggestion
## Next steps

On top of querying sensor data with third-party tools, you can also [query it with the Python SDK](/use-cases/sensor-data-query-sdk/) or [visualize it](/use-cases/sensor-data-visualize/).

{{< cards >}}
{{% card link="/use-cases/sensor-data-query-sdk/" %}}
{{% card link="/use-cases/sensor-data-visualize/" %}}
{{< /cards >}}

To see sensor data in action, check out these tutorials:

{{< cards >}}
{{% card link="/tutorials/control/air-quality-fleet/" %}}
{{< /cards >}}
```

---
title: "Query sensor data with third-party tools"
linkTitle: "Query sensor data"
weight: 30
type: "docs"
images: ["/services/icons/data-query.svg"]
description: "Query sensor data that you have synced to the Viam app using the data management service with SQL or MQL."
modulescript: true
aliases:
  - /manage/data/query/
  - /data/query/
# SME: Devin Hilly
---

You can use the data management service to capture sensor or time-series data from any machine and sync that data to the cloud.
Then, you can query it using {{< glossary_tooltip term_id="sql" text="SQL" >}} or {{< glossary_tooltip term_id="mql" text="MQL" >}} to obtain actionable insights or connect it to third-party visualization tools.

For example, you can configure data capture for several sensors on one machine, or for serveral sensors across multiple machines, to report the ambient operating temperature.
You can then run queries against that data to search for outliers or edge cases, to analyze how the ambient temperature affects your machines' operation.

> {{< alert title="Content" color="tip" >}}

Author: ```suggestion
{{< alert title="In this page" color="tip" >}}
```

> 
> This guide will walk you through:

Author: ```suggestion
```

---
title: "Query sensor data with third-party tools"
linkTitle: "Query sensor data"
weight: 30
type: "docs"
images: ["/services/icons/data-query.svg"]
description: "Query sensor data that you have synced to the Viam app using the data management service with SQL or MQL."
modulescript: true
aliases:
  - /manage/data/query/
  - /data/query/
# SME: Devin Hilly
---

> You can use the data management service to capture sensor or time-series data from any machine and sync that data to the cloud.
> Then, you can query it using {{< glossary_tooltip term_id="sql" text="SQL" >}} or {{< glossary_tooltip term_id="mql" text="MQL" >}} to obtain actionable insights or connect it to third-party visualization tools.
> 
> For example, you can configure data capture for several sensors on one machine, or for serveral sensors across multiple machines, to report the ambient operating temperature.
> You can then run queries against that data to search for outliers or edge cases, to analyze how the ambient temperature affects your machines' operation.

Reviewer: ```suggestion
You can use the data management service to [capture sensor data](/use-cases/sensor-data-query/) from any machine and sync that data to the cloud.
Then, as we'll cover on this page, you can query it using {{< glossary_tooltip term_id="sql" text="SQL" >}} or {{< glossary_tooltip term_id="mql" text="MQL" >}} to obtain actionable insights.

For example, you can configure data capture for several sensors on one machine, or for several sensors across multiple machines, to report the ambient operating temperature.
You can then run queries against that data to search for outliers or edge cases, to analyze how the ambient temperature affects your machines' operation.
```

{{< alert title="In this page" color="tip" >}}

> 1. [Gathering data on any machine and syncing it to the cloud](#gather-and-sync-data).
> 1. [Configuring data query](#configure-data-query).
> 1. [Querying data from third-party tools](#query-data-using-third-party-tools).

Reviewer: ```suggestion
1. [Gather data on any machine and syncing it to the cloud](#gather-and-sync-data).
1. [Configure data query](#configure-data-query).
1. [Query data from third-party tools](#query-data-using-third-party-tools).
```
[nit] for consistency with the docs in general

{{< /alert >}}

> ## Prerequisites

Reviewer: Why not make the capture data use case a prereq for all three of these?

Author: I thought it would be nice to just include it as an include. But yeah can do

{{% expand "A running machine connected to the Viam app. Click to see instructions." %}}

{{% snippet "setup.md" %}}

{{% /expand%}}

{{% expand "At least one configured sensor. Click to see instructions." %}}

Navigate to the **CONFIGURE** tab of your machine's page in [the Viam app](https://app.viam.com).
Click the **+** icon next to your machine part in the left-hand menu and select **Component**.
Then [find and add a sensor model](/components/sensor/) that supports your sensor.

{{% /expand%}}

{{% expand "The Viam CLI to set up data query. Click to see instructions." %}}

{{< readfile "/static/include/how-to/install-cli.md" >}}

{{% /expand%}}

{{% expand "`mongosh` or another third-party tool for querying data. Click to see instructions." %}}

> You can [download the `mongosh` shell](https://www.mongodb.com/try/download/shell) to follow along.

Author: yeah it's optional, they could use any third-party tool. Clarified a little

For example, use the following query to search by a date range in the `mongosh` shell, using the JavaScript `Date()` constructor to specify an explicit start timestamp, and use the current time as the end timestamp:

```mongodb {class="command-line" data-prompt=">"}
// Switch to sensorData database:
use sensorData

// Set desired start and end times:
const startTime = new Date('2024-02-10T19:45:07.000Z')
const endTime = new Date()

// Run query using $match:
db.readings.aggregate(
    [
        { $match: {
            time_received: {
                $gte: startTime,
                $lte: endTime }
        } }
    ] )
```

{{% /expand%}}

## Next steps

On top of querying sensor data with third-party tools, you can also [query it with the Python SDK](/use-cases/sensor-data-query-sdk/) or [visualize it](/use-cases/sensor-data-visualize/).

{{< cards >}}
{{% card link="/use-cases/sensor-data-query-sdk/" %}}
{{% card link="/use-cases/sensor-data-visualize/" %}}
{{< /cards >}}

> To see sensor data in action, check out these tutorials:

Author: ```suggestion
To see sensor data in action, check out this tutorial:
```

---
title: "Query sensor data with third-party tools"
linkTitle: "Query sensor data"
weight: 30
type: "docs"
images: ["/services/icons/data-query.svg"]

> description: "Query sensor data that you have synced to the Viam app using the data management service with SQL or MQL."

Author: ```suggestion
description: "Query sensor data that you have synced to the Viam app using the Viam app with SQL or MQL."
```

linkTitle: "Collect and query sensor data"
weight: 30
type: "docs"
images: ["/services/icons/data-query.svg"]
description: "Query sensor data that you have synced to the Viam app using the data management service with SQL or MQL."
modulescript: true
aliases:
  - /manage/data/query/
  - /data/query/
# SME: Devin Hilly
---

You can use the data management service to capture sensor or time-series data from any machine and sync that data to the cloud.
Then, you can query it using {{< glossary_tooltip term_id="sql" text="SQL" >}} or {{< glossary_tooltip term_id="mql" text="MQL" >}} to obtain actionable insights or connect it to third-party visualization tools.

For example, you can configure data capture for several sensors on one machine, or for serveral sensors across multiple machines, to report the ambient operating temperature.
You can then run queries against that data to search for outliers or edge cases, to analyze how the ambient temperature affects your machines' operation.

{{< alert title="Content" color="tip" >}}

This guide will walk you through:

1. Gathering data on any machine and syncing it to the cloud.
1. Querying data in the Viam app.

> 1. Configuring data query

Reviewer 2: ```suggestion
1. Configuring data query.
```

```

{{% /tab %}}
{{< /tabs >}}
{{< /tablestep >}}
{{< tablestep >}}
**3. Review results**

Click **Run query** when ready to perform your query and get matching results.
Query results are displayed as a [JSON array](https://json-schema.org/understanding-json-schema/reference/array) below your query.

{{% expand "See examples" %}}

- The following shows a SQL query that filters by component name and specific column names, and its returned results:

  {{< imgproc src="/services/data/query-ui-results.png" alt="Viam App Data Query tab with a SQL query shown and results shown below including two matching records" resize="800x" >}}

- The following shows a SQL query that returns a count of records matching the search criteria:

  {{< imgproc src="/services/data/query-ui-numreadings.png" alt="Viam App Data Query tab with a SQL query shown with the resulting count of matching records displayed below" resize="800x" >}}

For more information on MQL syntax, see the [MQL (MongoDB Query Language)](https://www.mongodb.com/docs/manual/tutorial/query-documents/) documentation.

{{% /expand%}}

{{< /tablestep >}}
{{< /table >}}

## Configure data query

> {{< readfile "/static/include/how-to/query-data.md" >}}

Reviewer: Step 3 isn't displaying correctly unless the screen is very very wide

## Query data using third-party tools

You can use third-party tools, such as the [`mongosh` shell](https://www.mongodb.com/docs/mongodb-shell/), [MongoDB Compass](https://www.mongodb.com/docs/compass/current/), to query against both captured sensor and time-series data, as well as its metadata, including machine ID, organization ID, and [tags](/services/data/dataset/#image-tags).

{{< table >}}
{{< tablestep link="/use-cases/sensor-data-query/#configure-data-query">}}
**1. Connect to your Viam organization's data**

> Run the following command to connect to the Viam organization's MongoDB Atlas instance from `mongosh` using the connection URI you obtained during query configuration:

Reviewer 2: ```suggestion
Run the following command to connect to your Viam organization's MongoDB Atlas instance from `mongosh` using the connection URI you obtained during query configuration:
```
(opt)

```sh {class="command-line" data-prompt=">"}
mongosh "<YOUR-DB-CONNECTION-URI>"
```

{{< /tablestep >}}
{{< tablestep >}}
**2. Query data from a compatible client**

> Once connected, you can run SQL or MQL to query captured data directly.

Reviewer 2: ```suggestion
Once connected, you can run SQL or MQL commands to query captured data directly.
```
(opt) I think

Author: I think it's referred to as statements
```suggestion
Once connected, you can run SQL or MQL statements to query captured data directly.
```

> The following query search the `sensorData` database and `readings` collection, and get sensor readings from an ultrasonic sensor on a specific `robot_id` where the recorded `distance` measurement is greater than `.2` meters.

Reviewer 2: ```suggestion
The following query searches the `sensorData` database and `readings` collection, and gets sensor readings from an ultrasonic sensor on a specific `robot_id` where the recorded `distance` measurement is greater than `.2` meters.
```

For example, to query against a field named `value` which is a subfield of the `data` field in the `readings` collection, you would use:

```mongodb {class="command-line" data-prompt=">"}
select data.`value` from readings
```

See the [MongoDB Atlas Documentation](https://www.mongodb.com/docs/atlas/data-federation/query/sql/language-reference/#compatability-and-limitations) for more information.

{{< /alert >}}

For information on connecting to your Atlas instance from other MQL clients, see the MongoDB Atlas [Connect to your Cluster Tutorial](https://www.mongodb.com/docs/atlas/tutorial/connect-to-your-cluster/).

<!-- markdownlint-disable-file MD001 -->

{{% expand "Need to query by date? Click here." %}}

##### Query by date

When using MQL to query your data by date or time range, you can optimize query performance by avoiding the MongoDB `$toDate` expression, using the [BSON `date` type](https://www.mongodb.com/docs/manual/reference/bson-types/#date) instead.

> For example, you could use the following query to search by a date range in the `mongosh` shell, using the JavaScript `Date()` constructor to specify an explicit start timestamp, and use the current time as the end timestamp:

Reviewer 2: ```suggestion
For example, use the following query to search by a date range in the `mongosh` shell, using the JavaScript `Date()` constructor to specify an explicit start timestamp, and use the current time as the end timestamp:
```
or "you can"? Not sure but conditional seems slightly off. Maybe opt

```mongodb {class="command-line" data-prompt=">"}
// Switch to sensorData database:
use sensorData

// Set desired start and end times:
const startTime = new Date('2024-02-10T19:45:07.000Z')
const endTime = new Date()

// Run query using $match:
db.readings.aggregate(
    [
        { $match: {
            time_received: {
                $gte: startTime,
                $lte: endTime }
        } }
    ] )
```

{{% /expand%}}

## Search data with the Python SDK

{{< table >}}
{{< tablestep link="/build/program/#requirements">}}
**1. Install the Python SDK**

For macOS (both Intel x86_64 and Apple Silicon) or Linux (x86, aarch64, armv6l), run the following commands:

```sh {class="command-line" data-prompt="$"}
python3 -m venv .venv
source .venv/bin/activate
pip install viam-sdk
```

{{< /tablestep >}}
{{< tablestep link="/cli/#organizations">}}
**2. Create an API key**

To access your machines using the Python SDK, you must use an API key:

```sh {class="command-line" data-prompt="$"}
viam organizations api-key create --org-id <org-id> --name my-api-key
```

{{< /tablestep >}}
{{< tablestep link="/appendix/apis/data-client/">}}

> **3. Use the API key with the `data_client`**

Reviewer 2: I think this could use more instructions about what an api key / key ID is and how to get them from app.viam.com -- I guess create command returns this so might not be necessary but maybe just a sentence about what it is/link to permissions page?
